internal/db/fetcher/prefix.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package fetcher
+
+import (
+	"context"
+	"slices"
+	"strings"
+
+	dsq "github.com/ipfs/go-datastore/query"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/datastore/iterable"
+	"github.com/sourcenetwork/defradb/internal/keys"
+)
+
+// prefix is a fetcher type responsible for iterating through multiple prefixes.
+//
+// It manages the document fetcher instances that will do the actual scanning.
+type prefix struct {
+	// The prefixes that this prefix fetcher must fetch from.
+	prefixes []keys.DataStoreKey
+	// The Iterator this prefix fetcher will use to scan.
+	kvIter iterable.Iterator
+
+	// The index of the current prefix being fetched.
+	currentPrefix int
+	// The child document fetcher, specific to the current prefix.
+	fetcher *document
+
+	// The below properties are only held here in order to pass them on to the next
+	// child fetcher instance.
+	ctx        context.Context
+	fieldsByID map[uint32]client.FieldDefinition
+	status     client.DocumentStatus
+	execInfo   *ExecInfo
+}
+
+var _ fetcher = (*prefix)(nil)
+
+func newPrefixFetcher(
+	ctx context.Context,
+	txn datastore.Txn,
+	prefixes []keys.DataStoreKey,
+	col client.Collection,
+	fieldsByID map[uint32]client.FieldDefinition,
+	status client.DocumentStatus,
+	execInfo *ExecInfo,
+) (*prefix, error) {
+	kvIter, err := txn.Datastore().GetIterator(dsq.Query{})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(prefixes) == 0 {
+		// If no prefixes are provided, scan the entire collection.
+		prefixes = append(prefixes, keys.DataStoreKey{
+			CollectionRootID: col.Description().RootID,
+		})
+	} else {
+		uniquePrefixes := make(map[keys.DataStoreKey]struct{}, len(prefixes))
+		for _, prefix := range prefixes {
+			// Deduplicate the prefixes to make sure that any given document is only yielded
+			// once.
+			uniquePrefixes[prefix] = struct{}{}
+		}
+
+		prefixes = make([]keys.DataStoreKey, 0, len(uniquePrefixes))
+		for prefix := range uniquePrefixes {
+			prefixes = append(prefixes, prefix)
+		}
+
+		// Sort the prefixes, so that documents are returned in the order they would be if the
+		// whole store was scanned.
+		slices.SortFunc(prefixes, func(a, b keys.DataStoreKey) int {
+			return strings.Compare(a.ToString(), b.ToString())
+		})
```

Comment:
suggestion: I assume the array of prefixes is not going to be large for any reasonable case and this approach would be fine.

The `ToString()` is not the cheapest and will be called 2 N (log N) time. We could make it just N.

The sorting itself could be also made using radix-like sorting style where we first sort by `CollectionRootID`, then by `InstanceType`... This won't give us exact scanner's sequence though.

Reply:
I don't think this is worth the maintenance cost atm, whilst probably more significant the the `make` comment, it's user-impact is still unmeasured, and IMO we have far better things to spend our time on atm.